assembly/src/assembler/mod.rs

```diff
@@ -62,18 +60,31 @@ pub struct Assembler {
 // ------------------------------------------------------------------------------------------------
 /// Constructors
 impl Assembler {
-    /// Start building an [Assembler]
-    pub fn new() -> Self {
-        Self::default()
-    }
+    /// Start building an [`Assembler`] with a kernel defined by the provided [CompiledLibrary].
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The library does not contain a kernel module.
+    /// - The library contains non-kernel modules.
+    pub fn with_kernel(kernel_lib: CompiledLibrary) -> Result<Self, AssemblerError> {
+        // get the kernel module from the library and make sure it is valid
+        let kernel_module = {
+            let modules = kernel_lib.into_module_infos().collect::<Vec<_>>();
+            if modules.is_empty() {
+                return Err(AssemblerError::EmptyKernelLibrary);
+            }
 
-    /// Start building an [`Assembler`] with the given [`Kernel`].
-    pub fn with_kernel(kernel: Kernel) -> Self {
-        let mut assembler = Self::new();
+            if modules.len() > 1 {
+                return Err(AssemblerError::NonKernelModulesInKernelLibrary);
+            }
 
-        assembler.module_graph.set_kernel(None, kernel);
+            modules.into_iter().next().expect("there must be exactly one module")
```

Comment:
This feels to me like the concept of "kernel library" doesn't fit well into the concept of "library".

We could create a new type `KernelLibrary`, which only stores a single `ModuleInfo`. And add `TryFrom<CompiledLibrary> for KernelLibrary` and `From<KernelLibrary> for CompiledLibrary` conversions.

Reply:
+1 to having a newtype for any `CompiledLibrary` that meets the kernel criteria by construction.

In general though, I'm not sure we actually need this restriction to be honest (no more than a single module). The only real restriction we should impose is that the library only exports kernel procedures. When resolving `syscall` targets in the `ModuleGraph`, we can simply limit our search to look only at exports from modules in the kernel library.

That said, there are a couple of things we need to address in order to support multi-module kernels:

* Kernel procedures are treated as being in a special global namespace, i.e. a procedure exported with the same name from two different kernel modules should not be allowed (otherwise `syscall.foo`, where `foo` has two different definitions in the kernel library, cannot be resolved). This validation would need to be added, but is simple to do so when assembling a kernel.
* We will need to update places in semantic analysis/the module graph, where we assume that the kernel is a single module.
* We need to figure out how to share procedures between modules in the kernel library itself. Currently, kernel modules can only export procedures that can be `syscall`'d, nothing else. I think the solution to this is to introduce a new visibility type (e.g. "protected" or something to that effect), which can be used to indicate that a procedure should only be resolvable by modules in the same library.

Reply:
I've added a `KernelLibrary` struct. For now, we assume that `KernelLibrary` is built from a single module. Totally agree that we should move towards multi-module kernel libraries, but this can be done in future PRs.

> We need to figure out how to share procedures between modules in the kernel library itself. Currently, kernel modules can only export procedures that can be `syscall`'d, nothing else. I think the solution to this is to introduce a new visibility type (e.g. "protected" or something to that effect), which can be used to indicate that a procedure should only be resolvable by modules in the same library.

Another option to handle this is via annotation.